### ISSUE TITLE
Preserve Special Tokens When CharTokenizer Truncates Input

### ISSUE THREAD

**Issue by aaishwarymishra** [pytorch-ignite/trainite#143]
## Description

When `CharTokenizer.__call__` is used with `add_special_tokens=True`, `truncation=True`, and `max_length`, it currently builds `[BOS] + tokens + [EOS]` and then keeps the first `max_length` IDs. Long inputs therefore lose the EOS token.

Example:

```python
tokenizer = CharTokenizer()
result = tokenizer("abcdef", truncation=True, max_length=5)
```

The result should fit within five tokens while retaining both BOS and EOS.

## Task

Update truncation in `trainite/preprocessors/char_tokenizer.py` so that:

- Output never exceeds `max_length`.
- BOS and EOS are retained when `add_special_tokens=True` and `max_length` has room for both.
- Behavior without special tokens remains unchanged.
- The attention mask still matches the returned IDs.

Add focused tests in `tests/preprocessors/char_test.py`.

## Acceptance criteria

- Add tests for truncation with and without special tokens.
- Define sensible behavior for `max_length` values smaller than two and test it.
- Do not add dependencies or tokenizer abstractions.
- `uv run pytest` passes.

**Comment by cananoo** [comment]
I plan to fix truncation by reserving space for BOS/EOS when special tokens are enabled, while preserving the existing no-special-token behavior and defining the max_length < 2 boundary. I will add focused tests for the special-token, no-special-token, and short-limit cases.

**Comment by aaishwarymishra** [comment]
Hi, we really appreciate your help, but this issue is for an event sorry :)

**Comment by BaselHamdy1** [comment]
I’m participating in CATReloaded 2026 and I’d like to work on this issue as part of the competition.
